grl/environment/pomdp_file.py

```diff
@@ -390,9 +405,96 @@ def __get_pi_phi(self, i):
 
         return i + 1
 
+    def convert_obs_actions(self):
+        """
+        If we're here, we want to convert our POMDP phi-action tensor
+        into a phi tensor (so they're all the same).
+
+        To do so, we add an extra "initialization" observation
+        We also need to expand our state space to include the last action.
+
+        ONE BIG ASSUMPTION: In the POMDPs with phi(s, a), the action
+        is applied to the state BEFORE an observation comes out.
+        """
+        # first we construct our new transition function.
+        # we go from |A| x |S| x |S| -> |A| x (|S| + 1)|A| x (|S| + 1)|A|
+        # states are ordered as s0a0, s0a1, s0a2, ..., s1a0, s1a1, ..., etc.
+        n_actions = self.T.shape[0]
+        og_n_states = self.T.shape[1]
+
+        # T_extra_start is |A| x (|S| + 1) x (|S| + 1)
+        start_expanded = np.expand_dims(np.expand_dims(self.start, 0), 0).repeat(n_actions, axis=0)
+        T_extra_start = np.concatenate((self.T, start_expanded), axis=1)
+        R_extra_start = np.concatenate((self.R, np.zeros_like(start_expanded)), axis=1)
+
+        cannot_transition_to_s0 = np.zeros((T_extra_start.shape[0], T_extra_start.shape[1], 1))
+        no_rewards_to_s0 = np.zeros_like(cannot_transition_to_s0)
+        T_extra_start = np.concatenate((T_extra_start, cannot_transition_to_s0), axis=2)
+        R_extra_start = np.concatenate((R_extra_start, no_rewards_to_s0), axis=2)
+
+        extra_n_states = og_n_states + 1
+
+        # Now we expand our T and R to add previous actions
+        # We start with the transition function
+        new_T = np.zeros((n_actions, extra_n_states * n_actions, extra_n_states * n_actions))
+        T_repeat_start_state = T_extra_start.repeat(n_actions, axis=1)
+        for i in range(new_T.shape[0]):
+            for j in range(new_T.shape[1]):
+                new_T[i, j, np.arange(extra_n_states) * n_actions + i] = T_repeat_start_state[i, j]
```

Comment:
The "preceding" action of the "next" state needs to match the selected action, so we copy these probabilities and leave the rest at zero.